[DEV-INTERNAL-API] Build TensorFlow Graphs from Keras

phi-dbq opened the pull request:
## What changes are proposed in this pull request?

We built a thin wrapper on top of TensorFlow's Session
to help us creating functions (and specifically, UDFs)
and composing subgraph (either Keras or TensorFlow).

This is functionality is suitable for **INTERNAL DEVELOPER ONLY**.

## How is this patch tested?

- [x] Unit tests
- [ ] Integration tests
- [x] Manual tests

phi-dbq committed:
Building UDFs from TensorFlow Graphs

We built a thin wrapper on top of TensorFlow's Session
to help us creating functions (and specifically, UDFs)
and composing subgraph (either Keras or TensorFlow).

phi-dbq committed:
Helpers from JVM's side

phi-dbq committed:
GraphFunction and GraphBuilderSession usage

phi-dbq committed:
(graph builder) unit tests

phi-dbq committed:
not exporting developer APIs

phi-dbq committed:
graph factory unit tests

phi-dbq committed:
update graph factory tests

phi-dbq commented on python/tests/graph_factory_test.py:
```diff
@@ -0,0 +1,159 @@
+            self.assertTrue(np.all(preds_tgt == preds_ref))
+
+    def test_pipeline(self):
+        """ Pipeline should provide correct function composition """
```
This is pretty much what's behind the SQL UDF.

sueann commented on python/sparkdl/graph_builder.py:
```diff
@@ -0,0 +1,329 @@
+        return (feeds, fetches)
+
+
+class GraphBuilderSession(object):
```
`IsolatedSession` might be a good name to describe what it's doing. It should match what tf.Session is (including being able to get the graph from the session), but ensure isolation for tensorflow and keras sessions.

sueann reviewed (commented):
As we talked offline, some overall suggestions:
* put the graph stuff in a module e.g. python/sparkdl/tfutils
* (Unfortunately) we use camelCase for function names in this package, following pyspark.

sueann commented on python/sparkdl/graph_builder.py:
```diff
@@ -0,0 +1,329 @@
+            ('input must be Placeholder, but get', op.type)
+        return name
+
+    def show_tf_graph(self, max_const_size=32):
```
ok so this writes out a html file so that you can view the graph in your browser. it's awesome! let's rename it to write_visualization_html() and have it take in an optional filename

phi-dbq committed:
PR comments: migrating utils to a separate module

1. Move JVMAPI to a submodule as it is more pythonic
2. Some unit tests on JVMAPI

phi-dbq commented on python/sparkdl/utils.py:
```diff
@@ -20,38 +21,60 @@
-    """
-    Loads the PythonInterface object (lazily, because the spark context needs to be initialized
-    first).
+class JVMAPI(object):
```
I am moving this into a separate submodule.

phi-dbq commented on python/tests/graph_builder_test.py:
```diff
@@ -0,0 +1,131 @@
+        self.assertEqual(str(gfn.graph_def), str(gdef_ref))
+
+
+    def test_get_graph_elemets(self):
```
oops ~

phi-dbq committed:
PR comments: graph utils refactoring

1. More meaningful names: `GraphBuilderSession` => `IsolatedSession`
2. Tensorflow graph utils resides in a single module
   (as opposed to a static class)
3. `GraphFunction` static methods

phi-dbq commented on python/sparkdl/graph_builder.py:
```diff
@@ -0,0 +1,329 @@
+
+logger = logging.getLogger('sparkdl')
+
+class GraphBuilder(object):
```
I moved the utility functions into a separate module and combined the `tf.Session` like features into a single class.

phi-dbq commented on python/sparkdl/graph_builder.py:
```diff
@@ -0,0 +1,329 @@
+
+    def __init__(self, graph):
+        assert isinstance(graph, tf.Graph)
+        self.sess = tf.Session(graph=graph)
```
I combined the `tf.Session` like features of `GraphBuilder` and `GraphBuilderSession` into a single class named `IsolatedSession`.

sueann commented on python/sparkdl/utils/graph_utils.py:
```diff
@@ -0,0 +1,163 @@
+#
```
why not put it in python/sparkdl/graph/utils.py ? I think it's better to put specialized util modules in the relevant submodules vs putting all util modules in the same "utils" module.

sueann commented on python/sparkdl/graph/graph_builder.py:
```diff
@@ -0,0 +1,200 @@
+import tensorflow as tf
+import tensorframes.core as tfrm
+
+from ..utils import jvmapi as JVMAPI
```
why not use absolute imports? seems more robust. cc @MrBago

sueann replied:
also this import doesn't seem used?

sueann commented on python/sparkdl/graph/graph_factory.py:
```diff
@@ -0,0 +1,136 @@
+
+    return gfn
+
+def merge(functions):
```
is it too messy to put in GraphFunction as e.g. `from_graph_functions` ? it's a little out of place in this module.

phi-dbq commented on python/sparkdl/graph_factory.py:
```diff
@@ -0,0 +1,168 @@
+                first_input_info.append((tnsr.dtype, tnsr.shape, name))
+
+        # Build a linear chain of all the provide functions
+        with GraphBuilderSession() as builder: 
```
In this session, the new first graph function is reloaded with its own name scope (to avoid potential name conflict). In the previous session, we acquire the tensor info of the original inputs. In this session, the original inputs are reconstructed.

sueann commented on python/sparkdl/graph/graph_factory.py:
```diff
@@ -0,0 +1,136 @@
+#
```
still woud like a better name for the module since "factory" sounds like https://en.wikipedia.org/wiki/Factory_method_pattern

sueann replied:
pieces.py ?

sueann commented on python/sparkdl/graph/graph_builder.py:
```diff
@@ -0,0 +1,200 @@
+#
```
could we name this file builder.py and if needed do 
```
import sparkdl.graph.builder as graph_builder
```
?

is there a convention @MrBago  ?

sueann commented on python/sparkdl/utils/jvmapi.py:
```diff
@@ -0,0 +1,51 @@
+#
```
is this module basically meant to be exposed only through `utils.pyutils` below?  might want to add a comment.

phi-dbq replied:
The `pyutils` is the one from spark's [PythonUtils](https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/api/python/PythonUtils.scala).

sueann commented on python/sparkdl/utils/utils.py:
```diff
@@ -0,0 +1,42 @@
+#
```
a little weird to have a module utils.utils

sueann reviewed (commented):
some high-level comments - will look at the implementation details in a bit. love the smaller modules!

phi-dbq committed:
PR comments

1. Renaming submodules
2. Using absolute import as per PEP 8

phi-dbq commented on python/sparkdl/graph/utils.py:
```diff
@@ -0,0 +1,163 @@
+
+logger = logging.getLogger('sparkdl')
+
+"""
```
Utilities defined in this module are supposed to run as if they were part of TensorFlow utilities, e.g.
`import sparkdl.graph.utils as tfx`.  Thus I kept the names as snake_case.

phi-dbq commented on python/sparkdl/graph/builder.py:
```diff
@@ -0,0 +1,250 @@
+
+        :param fpath: str or path, path to the serialized GraphFunction
+        """
+        _st = {"graph_def_bytes": self.graph_def.SerializeToString(),
```
we might want to decide a better serialized format, so the scala side can also load it easily

phi-dbq commented on python/tests/graph/test_pieces.py:
```diff
@@ -0,0 +1,165 @@
+        img_fpaths = glob(os.path.join(_getSampleJPEGDir(), '*.jpg'))
+
+        xcpt_model = Xception(weights="imagenet")
+        stages = [('spimage', gfac.buildSpImageConverter(SparkMode.RGB_FLOAT32)),
```
The model is supposed to look like this 
![graph-function-from-list](https://user-images.githubusercontent.com/28275034/27061537-99313b82-4f99-11e7-9f55-d4127fbb2289.png)

MrBago commented on python/tests/graph/test_builder.py:
```diff
@@ -0,0 +1,133 @@
+            self.assertEqual("x", tfx.op_name(g, x))
+            self.assertEqual("z", tfx.op_name(g, z))
+            self.assertEqual(tfx.get_tensor(g, z), z)
+            self.assertEqual(tfx.get_tensor(g, x), x)
```
Aren't lines 78-79 the same as lines 86-87?

MrBago commented on python/tests/graph/test_builder.py:
```diff
@@ -0,0 +1,133 @@
+            self.assertEqual("x:0", tfx.tensor_name(g, x))
+            self.assertEqual(g.get_operation_by_name("x"), tfx.get_op(g, x))
+            self.assertEqual("x", tfx.op_name(g, x))
+            self.assertEqual("x", tfx.op_name(g, x))
```
Is there a reason you get "x" twice here?

phi-dbq replied:
looks like it was a copy-and-paste duplicate ~ removed

MrBago reviewed (commented):
This looks great @phi-dbq. I comments are mostly pretty minor. My biggest concern is the joblib dependency, I prefer to avoid that if we could.

sueann commented:
Aside from @MrBago 's comments, my main comment would be that graph.utils contain functions that are basically equivalent to those here: https://github.com/databricks/spark-deep-learning/blob/53e5f56aefc0b1766c1b3424814357deeca36d7b/python/sparkdl/transformers/utils.py#L40-L71, so we should replace the uses of the latter with the graph.utils functions. This can be done in another PR, unless you are blocked by this PR and the scala PR getting in and so it's easier to just do it in this PR.

phi-dbq committed:
PR comments:

1. Removed `joblib` dependency
2. Keras model import minor refactor

phi-dbq commented on python/sparkdl/graph/builder.py:
```diff
@@ -0,0 +1,250 @@
+        jl.dump(_st, fpath)
+
+    @classmethod
+    def fromFile(cls, fpath):
```
I renamed this to `fromSerialized`

phi-dbq commented on python/sparkdl/graph/builder.py:
```diff
@@ -0,0 +1,250 @@
+
+        :param fpath: str or path, path to the serialized GraphFunction
+        """
+        _st = jl.load(fpath)
```
renamed as well

MrBago commented on python/requirements.txt:
```diff
@@ -6,3 +6,4 @@ numpy==1.11.2
 pillow==4.1.1
 tensorflow==1.1.0
 pygments==2.2.0
+six=1.10.0
```
This should be `six==1.10.0`.

MrBago commented on python/sparkdl/graph/builder.py:
```diff
@@ -0,0 +1,251 @@
+# limitations under the License.
+#
+import logging
+from pathlib import Path
```
I think pathlib is python3 only, can you check.

MrBago commented on python/sparkdl/graph/builder.py:
```diff
@@ -0,0 +1,251 @@
+        if not fpath.endswith("gfn"):
+            fpath += ".gfn"
+        with open(fpath, 'wb') as fout:
+            pickle.dump(serialized, fout)
```
Why aren't we using json here? Isn't serialized made up of strings and lists? json is much better for portability than pickle.

MrBago commented on python/sparkdl/graph/pieces.py:
```diff
@@ -0,0 +1,79 @@
+
+    return gfn
+
+def buildIdentity():
```
If this is for testing can we move it into the test file.

MrBago reviewed (commented):
This PR is looking really good @phi-dbq. Just a few more minor comments.

MrBago commented on python/sparkdl/transformers/tf_image.py:
```diff
@@ -25,7 +25,7 @@
     keyword_only, HasInputCol, HasOutputCol, SparkDLTypeConverters)
 import sparkdl.transformers.utils as utils
-from sparkdl.utils import list_to_vector_udf
+import sparkdl.utils.udf
```
Travis found more more thing :). I think you meant to do `import sparkdl.utils.udf as udf` here because you use the module on L239. But that won't actually work because you're shadowing the udf from `pyspark.sql.functions` imported a few lines above. Maybe we should rename this module?

phi-dbq replied:
oh well ~ that was subtle TT. Moving them back to JVMAPI ~ thanks man!

phi-dbq committed:
PR comments

1. fix requirements.txt
2. removed pathlib in favor of os.path
3. removed identity piece
4. using JSON + binary file for graph function storage

MrBago commented on python/sparkdl/graph/builder.py:
```diff
@@ -0,0 +1,263 @@
+        gdef_bytes_fpath = "{}.gdef.bytes".format(fpath)
+        with open(gdef_bytes_fpath, 'wb') as fout:
+            gdef_bytes = self.graph_def.SerializeToString()
+            fout.write(gdef_bytes)
```
oic what the issue here is, even though the method is called `SerializeToString()` it returns a binary object ... that's lame. hmm ...

I'm not sure the best way to handle this. We could pickle ... or use 2 files ... or base64 encode the blob ... I don't love any of those options :).

Sorry for mouthing off without fully understanding the issue :p.

phi-dbq replied:
They certainly have interesting ways to name things. Thanks for the suggestions. I think saving them for two files like this makes Scala Python model sharing easier.

phi-dbq committed:
utils update

phi-dbq committed:
temporary travis

phi-dbq committed:
utils fix

phi-dbq commented on python/tests/tests.py:
```diff
@@ -33,7 +33,7 @@ class SparkDLTestCase(unittest.TestCase):
     @classmethod
     def setUpClass(cls):
-        cls.sc = SparkContext('local[4]', cls.__name__)
+        cls.sc = SparkContext('local[*]', cls.__name__)
```
To make Travis happy, allocating executors according to the machine's capacity.

thunterdb replied:
If we do not do that, does it take so long that we get killed? We should make the tests run faster somehow.

phi-dbq commented on python/sparkdl/graph/utils.py:
```diff
@@ -61,13 +61,25 @@ def get_tensor(graph, tfobj_or_name):
+        'cannot locate tensor {} in current graph'.format(_tensor_name)
+    return tnsr
+
+def as_tensor_name(name):
```
This method does not check the whether said tensor name exists in the graph.

thunterdb replied:
Add this comment in the code

phi-dbq commented on python/run-tests.sh:
```diff
@@ -93,11 +93,19 @@ else
-  $PYSPARK_DRIVER_PYTHON -m "nose" -v --exe "$noseOptions" 2>&1 | grep -vE "INFO (ParquetOutputFormat|SparkContext|ContextCleaner|ShuffleBlockFetcherIterator|MapOutputTrackerMaster|TaskSetManager|Executor|MemoryStore|CacheManager|BlockManager|DAGScheduler|PythonRDD|TaskSchedulerImpl|ZippedPartitionsRDD2)";
+  $PYSPARK_DRIVER_PYTHON \
+      -m "nose" \
+      --nocapture --nologcapture \
```
To avoid overly being clogged by py4j output.

MrBago replied:
I think --nologcapture will really help. Can we avoid using --nocapture, nocapture allows us to put useful print statements in the tests which will only get reported if the test fails.

MrBago commented on python/sparkdl/graph/builder.py:
```diff
@@ -0,0 +1,263 @@
+            model_path = os.path.join(mkdtemp(prefix='kera-'), "model.h5")
+            # Save to tempdir and restore in a new session
+            model.save(str(model_path), overwrite=True)
+            gfn = load_model_file(str(model_path))
```
Doesn't `os.path.join` return a str?

MrBago commented on python/sparkdl/graph/builder.py:
```diff
@@ -0,0 +1,263 @@
+            # Save to tempdir and restore in a new session
+            model.save(str(model_path), overwrite=True)
+            gfn = load_model_file(str(model_path))
+            shutil.rmtree(os.path.dirname(model_path), ignore_errors=True)
```
I like to put everything between `mktemp` and `rmtree` in a try/finally :).

phi-dbq committed:
PR comments

MrBago commented on python/sparkdl/graph/builder.py:
```diff
@@ -0,0 +1,264 @@
+        if isinstance(model_or_file_path, KerasModel):
+            model = model_or_file_path
+            try:  # Save to tempdir and restore in a new session
+                model_path = os.path.join(mkdtemp(prefix='kera-'), "model.h5")
```
`mkdtemp` should not be inside the try because we do not want to call `rmtree` if `mkdtemp` fails.

phi-dbq committed:
try catch tmpdir

thunterdb commented on python/sparkdl/transformers/tf_image.py:
```diff
@@ -236,6 +239,6 @@ def _convertOutputToVector(self, df, tfs_output_col):
         """
         return (
-            df.withColumn(self.getOutputCol(), list_to_vector_udf(df[tfs_output_col]))
+            df.withColumn(self.getOutputCol(), JVMAPI.list_to_vector(df[tfs_output_col]))
```
as mentioned in #15 , let's change that so that the python list is handlded directly by py4j.

thunterdb commented on python/sparkdl/utils/jvmapi.py:
```diff
@@ -0,0 +1,73 @@
+    """ Default JVM Python Interface class """
+    return forClass(javaClassName=PYTHON_INTERFACE_CLASSNAME)
+
+def list_to_vector(col):
```
we should not need that

thunterdb commented on python/sparkdl/utils/jvmapi.py:
```diff
@@ -0,0 +1,73 @@
+
+logger = logging.getLogger('sparkdl')
+
+def _curr_sql_ctx(sqlCtx=None):
```
this is a duplicate from some code in #15 ?

thunterdb commented on python/tests/graph/test_pieces.py:
```diff
@@ -0,0 +1,169 @@
+        img_fpaths = glob(os.path.join(_getSampleJPEGDir(), '*.jpg'))
+
+        xcpt_model = Xception(weights="imagenet")
+        stages = [('spimage', gfac.buildSpImageConverter(SparkMode.RGB_FLOAT32)),
```
Nice!

thunterdb commented on python/sparkdl/graph/utils.py:
```diff
@@ -0,0 +1,185 @@
+
+def get_op(graph, tfobj_or_name):
+    """ Get the graph element `Operation` from a given name or graph element """
+    if isinstance(tfobj_or_name, tf.Operation):
```
can you also add a check that the graph of `tfobj_or_name` is `graph`?

phi-dbq replied:
with a hard coded `local[4]`, travis get OOM for tests with inception graphs.

thunterdb reviewed (commented):
I just have a few quick comments.

phi-dbq commented on python/sparkdl/graph/graph_builder.py:
```diff
@@ -0,0 +1,252 @@
+
+    def dump(self, fpath):
+        """
+        Store the GraphFunction to a file
```
I actually changed this to explicitly store the `GraphDef` as bytes and the filename, inputs and outputs names in a JSON file. 
https://github.com/databricks/spark-deep-learning/blob/1dda1ef8c9d7ff561ee9bc85cbcddf9e4095466b/python/sparkdl/graph/builder.py#L140

phi-dbq committed:
PR comments

1. docs
2. graph type check in graph.utils

phi-dbq committed:
sbt build spill out classpath for testing

phi-dbq committed:
Revert "sbt build spill out classpath for testing"

This reverts commit 962e23aa4876cd26558c99e37c0970792e09c74a.

thunterdb commented on python/sparkdl/graph/pieces.py:
```diff
@@ -0,0 +1,71 @@
+
+    return gfn
+
+def buildFlattener():
```
one line doc with what is being returned?

thunterdb commented on python/sparkdl/graph/utils.py:
```diff
@@ -0,0 +1,255 @@
+    _shape = get_tensor(graph, tfobj_or_name).get_shape().as_list()
+    return [-1 if x is None else x for x in _shape]
+
+def get_op(graph, tfobj_or_name):
```
I am surprised we do not have this already.

thunterdb commented on python/sparkdl/graph/utils.py:
```diff
@@ -0,0 +1,255 @@
+    """
+    assert isinstance(name, six.string_types)
+    name_parts = name.split(":")
+    assert len(name_parts) <= 2
```
`, name_parts`

phi-dbq replied:
Oh?

thunterdb replied:
` assert len(name_parts) <= 2, name_parts` to print what is wrong when the assert gets triggered

thunterdb replied:
You should do that for any assert, otherwise it is unhelpful.

thunterdb commented on python/sparkdl/graph/utils.py:
```diff
@@ -0,0 +1,255 @@
+    """
+    assert isinstance(name, six.string_types)
+    name_parts = name.split(":")
+    assert len(name_parts) <= 2
```
`, name_parts`

thunterdb commented on python/sparkdl/graph/utils.py:
```diff
@@ -0,0 +1,255 @@
+
+def strip_and_freeze_until(fetches, graph, sess=None, return_graph=False):
+    """
+    Converting all variables into constants
```
you should mention that it prunes out all the parts of the graph that are unreachable from the fetches.

thunterdb commented on python/sparkdl/graph/utils.py:
```diff
@@ -0,0 +1,255 @@
+    else:
+        return gdef_frozen
+
+def write_visualization_html(graph, html_file_path=None, max_const_size=32, show_in_browser=True):
```
this is very cool! I wonder if we should advertise it more. Certainly worth of its own PR.

phi-dbq replied:
Thanks :) 
I am removing this and putting it to a separate PR.

thunterdb reviewed (commented):
@phi-dbq thanks a lot for taking over this code. I have a number of comments, but they are not very big. We should be able to merge it after that.

@sueann also wanted to take a look after that.

phi-dbq commented on .travis.yml:
```diff
@@ -1,10 +1,10 @@
 
 dist: trusty
 
+jdk:
```
I am mainly following [tensorframes' travis file](https://github.com/databricks/tensorframes/blob/master/.travis.yml#L6). 
Otherwise we get the minor version error as the tensorframes dependency we use is compiled against java 8

MrBago replied:
This is pretty frustrating. There is a separate PR for this, https://github.com/databricks/spark-deep-learning/pull/20. Adding travis related stuff to your PR not only introduces merge conflicts, which then take time to resolve. It also makes your PR harder to review and we end up having the same conversations twice.

Can we commit to these conditions for merging this PR. 1) Fix failing tests in master 2) Merge travis PR 3) rebase on master, then we can properly merge this PR.

phi-dbq commented on python/run-tests.sh:
```diff
@@ -93,11 +93,19 @@ else
 fi
 
+# Limit TensorFlow error message
+# https://github.com/tensorflow/tensorflow/issues/1258
```
I am not quite certain whether it is due to `--nologcapture` or that plus this that suppressed the TensorFlow INFO logs. 

For tests, I believe they did run
https://travis-ci.org/databricks/spark-deep-learning#L1099

thunterdb replied:
indeed, they did.

thunterdb replied:
it does, indeed.

phi-dbq commented on python/sparkdl/graph/builder.py:
```diff
@@ -0,0 +1,268 @@
+        tf.import_graph_def(gfn.graph_def,
+                            input_map=input_map,
+                            return_elements=gfn.output_names,
+                            name=name,
```
Per https://www.tensorflow.org/api_docs/python/tf/import_graph_def, passing `None` is the default of `tf.import_graph_def` and it will add the prefix "import" to the graph.

phi-dbq committed:
addressing comments

fix tests for name changes

sueann commented on python/sparkdl/transformers/utils.py:
```diff
@@ -18,6 +18,7 @@
 from pyspark.ml.param import TypeConverters
 
 from sparkdl.image.imageIO import imageType
+import sparkdl.graph.utils as tfx
```
unused?

sueann commented on python/sparkdl/graph/builder.py:
```diff
@@ -0,0 +1,268 @@
+    """
+    Building TensorFlow graph and export as either UDF or GraphFunction
+
+    It provides a GraphBuilder object which provides
```
no longer true

sueann commented on python/tests/graph/test_builder.py:
```diff
@@ -0,0 +1,157 @@
+from ..transformers.image_utils import _getSampleJPEGDir, getSampleImagePathsDF
+
+
+class GraphBuilderTest(SparkDLTestCase):
```
there is no longer a class called GraphBuilder, so rename to... `GraphFunctionWithIsolatedSessionTest` ?

several of the tests are testing GraphFunction and nearly all are testing IsolatedSession. It'd be ideal to separate them out, e.g. testing GraphFunction functions should use tf.Session. But the test coverage looks good enough so I'd add a TODO comment for this and call it a day for this PR (it's been modified enough...)

sueann reviewed (commented):
almost there! a few more small comments.

phi-dbq committed:
further PR comments

phi-dbq committed:
remove dump

thunterdb commented on .travis.yml:
```diff
@@ -27,7 +27,6 @@ install:
 script:
   - ./build/sbt -Dspark.version=$SPARK_VERSION -Dscala.version=$SCALA_VERSION "set test in assembly := {}" assembly
   - ./build/sbt -Dspark.version=$SPARK_VERSION -Dscala.version=$SCALA_VERSION coverage test coverageReport
-  - SPARK_HOME=$HOME/.cache/spark-versions/$SPARK_BUILD ./python/run-tests.sh
```
@phi-dbq you just removed the python tests, see the travis exit...

phi-dbq committed:
revive python tests

phi-dbq committed:
assert message

sueann commented:
lgtm. going to merge... in 3, 2, 1.
